Rewrite mailers chapter (#120)
infinum/rails-handbook: Development practices/Emails.md

# Guidelines

## Templates

> In most cases we don't need a custom design for the emails. We use [simple HTML templates](https://github.com/mailgun/transactional-email-templates) and modify them slightly with project's theme colors, and add a logo if needed.

Author: Slicing a basic custom email is not a problem, but the design and multiple iterations with the client are often very expensive compared to the value for the client.
But it makes sense to develop a few basic and clean templates which we could use instead of those I pasted in the chapter, as those look outdated. I'll open a task so we can discuss this.

> If we will need a fully custom template, consult with someone from the JS team on which libraries to use.

Reviewer: ```suggestion
If there's a requirement for having a completely custom template consult with someone from the JS team on which libraries to use.
```

> If there's a requirement for having a completely custom template consult with someone from the JS team on which libraries to use.

Reviewer 2: Is there anything else in use except MJML? I knew it before, but the frontend developers didn't on a recent project, so it might make sense to list the most used options here so we know what to choose from.

Author: I'll mention MJML.

> If somebody from the other team will be working on emails, please make sure you first set up the Letter opener, prepare the previews for all mailers and instruct your colleague on how to use the [mailer previews](#mailer-previews).

Reviewer: ```suggestion
If somebody from the other team works on emails, please make sure you first set up the Letter opener, prepare the previews for all mailers and instruct your colleague on how to use the [mailer previews](#mailer-previews).
```

Present simple feels better to me here.. @cukoreva what do you think?

Reviewer 3: I definitely agree with not using a future tense, but maybe this would sound better:
```suggestion
If somebody from the other team is supposed to work on emails, please make sure you first set up the Letter opener, prepare the previews for all mailers and instruct your colleague on how to use the [mailer previews](#mailer-previews).
```
or
```suggestion
If somebody from the other team is expected to work on emails, please make sure you first set up the Letter opener, prepare the previews for all mailers and instruct your colleague on how to use the [mailer previews](#mailer-previews).
```

> If somebody from the other team is expected to work on emails, please make sure you first set up the Letter opener, prepare the previews for all mailers and instruct your colleague on how to use the [mailer previews](#mailer-previews).

Reviewer 2: If we're talking about MJML here theres a much simpler way
https://mjml.io/try-it-live
or 
```bash
mjml input.mjml -o output.html

```

Author: This is great for when you're slicing the template, but after that you still need to put the file to the repo, add some erb code and check how everything renders.

## Send emails asynchronously
Emails should be sent asynchronously from a background job:

> - so they don't block current HTTP request for longer than needed

Reviewer 3: ```suggestion
- so they don't block the current HTTP request longer than needed
```

> - if the email server is temporarily unavailable, the background job will fail and re-try again (check your re-try options!).

Reviewer: ```suggestion
- if the email server is temporarily unavailable, the background job will fail and re-try again (check your retry options!).
```

See the [Background jobs chapter](/books/rails/development-practices/background-and-scheduled-jobs) on how to configure those.

> If you're using Devise, don't forget to add and configure the [devise-async gem](https://github.com/mhfs/devise-async).

Reviewer 4: there is no need to use gem for this

Add this to user.rb:

```ruby
  def send_devise_notification(notification, *args)
    devise_mailer.send(notification, self, *args).deliver_later
  end
```

# Local development

We usually don't need nor want to send real emails to users in development, so we use [Letter opener](https://github.com/ryanb/letter_opener). Instead of sending an email, it will render the email in a new tab.

## Mailer previews
Rails has a built-in feature for [previewing emails](https://guides.rubyonrails.org/action_mailer_basics.html#previewing-emails). Using it allows you to see what the email looks like without sending a real email. Please read the setup instructions on the previous link.

> We typically add mailer preview classes to `spec/mailers/previews` folder.

Reviewer 3: ```suggestion
We typically add mailer preview classes to the `spec/mailers/previews` folder.
```

Here's an example of a preview class:
```ruby
# Preview all emails at http://localhost:3000/rails/mailers
class UserMailerPreview < ActionMailer::Preview
  def storage_limit_warning_email
    UserMailer.storage_limit_warning_email(user)
  end

  private

  def user
    User.new(email: 'rails@is.best', username: 'mockyMock', space_usage: 450)
  end
end
```

The important thing is not to load existing objects from your development database here, because you will have problems when you delete or change the data. Other developers working on the project probably won't have the same data in their database.
A better solution would be to just build the necessary objects with all attributes and relationships used in the template.

# Production and staging

In production/staging/uat we mainly use [Mailgun](https://mailgun.com) to send emails from our apps.

> Someone from the DevOps team should open a Mailgun account and set up everything for production/staging and store credentials in corresponding secrets files.

Reviewer: ```suggestion
Someone from the DevOps team should open a Mailgun account and set up everything for production/staging and store credentials in [Vault](../handling-secrets).
```

To set up Mailgun in the app, add the official [mailgun-ruby](https://github.com/mailgun/mailgun-ruby#rails) gem and set it up per environment as described in the gem's Readme.

Mailgun also has [webhooks](https://documentation.mailgun.com/en/latest/api-webhooks.html), if for an example you need to track failed deliveries due to having non-existing/non-verified emails in the database.

## Sending emails on behalf of users

> Sometimes you will need to send an email on behalf of the app's users, i.e. Contact form submissions. **Never set from field to user's email address - use Reply-To.**. Mailgun won't deliver emails with *from* address with a different domain than the one configured in Mailgun settings.

Reviewer: ```suggestion
Sometimes we need to send an email on behalf of the app's users, i.e. Contact form submissions. **Never set _from_ field to user's email address - use Reply-To.**. Mailgun won't deliver emails with _from_ address with a different domain than the one configured in Mailgun settings.
```

## Intercepting emails
Emails in non-production environments should have prefixed subjects, i.e. `[Staging] Confirmation Instructions`. Creating a custom interceptor is simple, for more info read [here](https://guides.rubyonrails.org/action_mailer_basics.html#intercepting-emails).

In some apps you should intercept emails in staging to prevent sending them to real users, or to avoid Mailgun failures if you have fake/non-existing emails in the database.

> There's a [mail_interceptor](https://github.com/bigbinary/mail_interceptor) gem which intercepts and forwards emails to a given address, and also sends the emails only to whitelisted addresses.

Reviewer 3: ```suggestion
There's the [mail_interceptor](https://github.com/bigbinary/mail_interceptor) gem which intercepts and forwards emails to a given address, and also sends the emails only to whitelisted addresses.
```
